atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2509,38 +2563,80 @@ private void logCommitLockTenureExceeded(
                 TransactionConstants.TRANSACTION_TABLE.getQualifiedName(),
                 TransactionConstants.getValueForTimestamp(getStartTimestamp())));
         cellCommitLocksRequested = cellLockCount;
         rowCommitLocksRequested = rowLockCount + 1;
-        return result;
+        return LocksAndMetadata.of(
+                lockDescriptors,
+                // For now, lock request metadata only consists of change metadata. If it is absent, we can save
+                // computation by not doing an index encoding
+                lockDescriptorToChangeMetadata.isEmpty()
+                        ? Optional.empty()
+                        : Optional.of(LockRequestMetadata.of(lockDescriptorToChangeMetadata)));
+    }
+
+    private void collectCellLocks(
+            Set<LockDescriptor> lockDescriptors,
+            Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata,
+            TableReference tableRef) {
+        Map<Cell, ChangeMetadata> changeMetadataForWrites = getChangeMetadataForWrites(tableRef);
+        for (Cell cell : getLocalWrites(tableRef).keySet()) {
+            LockDescriptor lockDescriptor =
+                    AtlasCellLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName(), cell.getColumnName());
+            lockDescriptors.add(lockDescriptor);
+            if (changeMetadataForWrites.containsKey(cell)) {
+                lockDescriptorToChangeMetadata.put(lockDescriptor, changeMetadataForWrites.get(cell));
+            }
+        }
+    }
+
+    private void collectRowLocks(
+            Set<LockDescriptor> lockDescriptors,
+            Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata,
+            TableReference tableRef) {
+        Map<Cell, ChangeMetadata> changeMetadataForWrites = getChangeMetadataForWrites(tableRef);
+        Cell lastCell = null;
+        for (Cell cell : getLocalWrites(tableRef).keySet()) {
+            LockDescriptor rowLockDescriptor =
+                    AtlasRowLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName());
+            if (changeMetadataForWrites.containsKey(cell)) {
+                if (lockDescriptorToChangeMetadata.containsKey(rowLockDescriptor)) {
+                    throw new SafeIllegalStateException(
+                            "Two different cells in the same row have metadata and we create locks on row level.",
+                            UnsafeArg.of("tableRef", tableRef),
+                            UnsafeArg.of("rowName", cell.getRowName()),
+                            UnsafeArg.of("existingMetadata", lockDescriptorToChangeMetadata.get(rowLockDescriptor)),
+                            UnsafeArg.of("newMetadata", changeMetadataForWrites.get(cell)));
+                }
+                lockDescriptorToChangeMetadata.put(rowLockDescriptor, changeMetadataForWrites.get(cell));
+            }
+            if (lastCell == null || !Arrays.equals(lastCell.getRowName(), cell.getRowName())) {
```

Comment:
So, I think this is an artifact of an older time, where we used to have to compute hash codes on every iteration, which made this fairly expensive. That said, this is no longer the case (their hashcode values are now cached), thus we should be OK with just doing `lockDescriptors.add(rowLockDescriptor)`